### ISSUE TITLE
Completion candidates for union types

### ISSUE THREAD

**Issue by alisnic** [castwide/solargraph#507]
Currently, if method signature includes multiple types, solargraph will provide candidates only for the first type. Heres a failing spec that can be added to `spec/source_map/clip_spec.rb`:

```ruby
  it 'complete all methods from union types' do
    source = Solargraph::Source.load_string(%(
      class Thing
        # @return [String, Array]
        def foo; end
      end
      Thing.new.foo.an
      Thing.new.foo.up
    ), 'test.rb')
    api_map = Solargraph::ApiMap.new.map(source)

    string_names = api_map.clip_at('test.rb', [6, 22]).complete.pins.map(&:name)
    expect(string_names).to eq(["upcase", "upcase!", "upto"])

    array_names = api_map.clip_at('test.rb', [5, 22]).complete.pins.map(&:name)
    expect(array_names).to eq(["any?"])
  end
```

Executing it, yields the following report:

```bash
Failures:

  1) Solargraph::SourceMap::Clip complete all methods from union types
     Failure/Error: expect(array_names).to eq(["any?"])

       expected: ["any?"]
            got: []
```

@castwide if you can give me some pointers on how to fix this, I would gladly contribute with a patch

**Comment by castwide** [comment]
This is a feature that's been on the radar for a while, but I still have some outstanding questions about the best approach.

* Should method candidates be a union (as in your test) or an intersection? In the latter, candidates would need to exist in both classes; for example, [Array, Set] would not include `Array#push` or `Set#add`, but it would include all `Enumerable` methods.
* If it uses an intersection, another useful feature would be awareness of code blocks that only apply to a specific class. Example:
  ```ruby
  # @param bar [Integer, String]
  def foo(bar)
    case bar
    when Integer
      bar.abs # <- OK, recognized as Integer
    when String
      bar.upcase # <- OK, recognized as String
      bar.abs # <- Error, String#abs not found
    else
      bar.class # <- OK, Object#class found in both Integer and String
      bar.abs # <- Error, String#abs not found
    end
  end
  ```

**Comment by castwide** [comment]
Good points. I think I'm somewhat conflating the requirements of two distinct features: the completion items provided by the language server, and the rules that apply to strict type checking. In the absence of strict type checking, a union of type methods makes sense.

I started to write an explanation of how it might be implemented, but it turned out to be just as fast to make it into a PR: https://github.com/castwide/solargraph/pull/512

**Comment by alisnic** [comment]
Awesome, thanks! let me know if I can be of any help in the MR

**Comment by castwide** [comment]
Released in v0.47.0.
